apps/src/applab/RecordListener.js

```diff
@@ -0,0 +1,204 @@
+    return false;
+  }
+
+  this.tableHandlers_[tableName] = new TableHandler(tableName, callback);
+
+  return true;
+};
+
+/**
+ * Clears the timeouts and resets internal state.
+ */
+RecordListener.prototype.reset = function () {
+  for (var tableName in this.tableHandlers_) {
+    this.tableHandlers_[tableName].reset();
+  }
+  this.tableHandlers_ = {};
+};
+
+//////////////////////////////////////////////////
+// TableHandler
+//////////////////////////////////////////////////
+
+/**
+ * This class encapsulates all the logic and state required for listening to
+ * changes to a single table.
+ * @param {string} tableName
+ * @param {function} callback
+ * @constructor
+ */
+var TableHandler = function (tableName, callback) {
+  /**
+   * @type {{string}}
+   * @private
+   */
+  this.tableName_ = tableName;
+
+  /**
+   *  @type {function}
+   *  @private
+   */
+  this.callback_ = callback;
+
+  /**
+   * @typedef {Object.<number, string>} IdToJsonMap Map from record id
+   * to a JSON-stringified copy of the corresponding record.
+   */
+
+  /**
+   * IdToJsonMap representing the contents of the table.
+   * We use this to efficiently test whether a record's contents has changed.
+   * @type {IdToJsonMap}
+   * @private
+   */
+  this.idToJsonMap_ = {};
+
+  /**
+   * Timeout id of the pending call to window.setTimeout.
+   * @type {number}
+   * @private
+   */
+  this.timeoutId_ = null;
+
+  this.scheduleNextFetch_();
+};
+
+/**
+ * @private
+ */
+TableHandler.prototype.scheduleNextFetch_ = function () {
+  this.timeoutId_ = setTimeout(function recordListener() {
+    this.fetchNewRecords_();
+  }.bind(this), RECORD_INTERVAL);
+};
+
+/**
+ * @private
+ */
+TableHandler.prototype.fetchNewRecords_ = function () {
+  var req = new XMLHttpRequest();
+  req.onreadystatechange = this.reportRecords_.bind(this, req);
+  var url = '/v3/shared-tables/' + Applab.channelId + '/' + this.tableName_;
+  req.open('GET', url, true);
+  req.send();
+};
+
+/**
+ * @param {XMLHttpRequest} req
+ * @private
+ */
+TableHandler.prototype.reportRecords_ = function (req) {
+  var done = XMLHttpRequest.DONE || 4;
+  if (req.readyState !== done) {
+    return;
+  }
+
+  if (req.status < 200 || req.status >= 300) {
+    // ignore errors
+    return;
+  }
+
+  if (!this.idToJsonMap_) {
+    // The maps may have been reset while we were fetching the data.
+    return;
+  }
+
+  // Performance notes: on a table with 2000 records on localhost it takes ~1.5 seconds
+  // to fetch the data from dynamodb and 30ms to process it here. Processing time may be
+  // longer on slower machines. One quick and dirty performance improvement is to cache
+  // the responseText here and return if it has not changed (only 1ms). However, a better
+  // performance improvement is to only poll when Pusher tells us to, which would render
+  // the responseText check unnecessary.
+
+  var callback = this.callback_;
+  var records = JSON.parse(req.responseText);
```

Comment:
I've moved the call to `scheduleNextFetch_` to before this line, so that an exception here doesn't stop polling indefinitely. Malformed JSON seems like it would represent a programming error here, so allowing this error to happen and show up in our monitoring tools seems right to me, but I am open to other ideas of how this should be handled.